### DISCUSSION TITLE
Session issues in production: /api/auth/session returns empty {} but works in development

### DISCUSSION THREAD

**Discussion by tayloraucoin** [nextauthjs/next-auth#11909] [Help]
### Environment

```
 System:
    OS: macOS 12.5
    CPU: (10) arm64 Apple M1 Pro
    Memory: 79.56 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.nvm/versions/node/v20.10.0/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 10.2.3 - ~/.nvm/versions/node/v20.10.0/bin/npm
  Browsers:
    Chrome: 128.0.6613.115
    Edge: 128.0.2739.63
    Safari: 15.6
  npmPackages:
    next: 14.2.5 => 14.2.5 
    next-auth: ^4.24.7 => 4.24.7 
    react: ^18 => 18.3.1 
```

### Reproduction URL

https://github.com/tayloraucoin/next-auth-prisma-undefined-bug

### Describe the issue

**This is the step-by-step flow of the feature that is causing issues:**
1. A user is generated through a checkout process.
2. A JWT-encoded token is generated for the user.
3. The user receives an email with a verification link that includes the token.
4. When they land on the sign-up page, the token is decoded.
5. A next-auth.session-token cookie is created and returned in the verify-token response, mimicking next-auth’s native sign-in behavior.
6. The session state is refreshed using `getSession`, which sends a verification request to next-auth.
7. The request to `/api/auth/session` is where the issue occurs in production—this endpoint returns `{}`, while it works as expected in development.

We are encountering session issues in **production** on **Vercel** with the Next.js app using next-auth. The issue arises during the `getSession` call, which fails to return a valid session. The session API request (`/api/auth/session`) is returning `{}`, and no session data is being retrieved for authenticated users, even though the cookies are properly set.

In development, everything works as expected, and the `getSession` call successfully retrieves the session data. However, in production, the same logic results in an empty object being returned, and the user remains unauthenticated.

We have used the `encode` and `decode` functions provided by next-auth to set a `next-auth.session-token` cookie in the browser after validating a token passed in a URL query string. This setup works as expected in development but fails in production.

I have tried adjusting cookie settings (e.g., SameSite, secure), but it does not resolve the issue. 

**Important: The callback functions with the nextauth setup are not being called in production.** _Only the `redirect` callback is_. However, in development, I can see the `jwt` and `session` callbacks are properly triggered. The absence of these callbacks is the biggest clue to what is going on. See `./app/api/auth/[...nextauth]/route.ts` for reference. 

### How to reproduce

Using the https://github.com/nextauthjs/next-auth-example template provided forces version 5. When trying to set next-auth to my actual version, several things broke. NextAuth from next-auth/next does not work in version 5. This has been frustrating, as the template is meant to save time. This example repo is also used for another issue with the Prisma adapter: https://github.com/nextauthjs/next-auth/issues/11773

In the setup, you can navigate to `http://localhost:3000/checkout-simulate`, enter an email, and click submit. This creates a user, generates an encoded JWT token, and produces a URL that redirects to `/sign-in?token=${token}`.

On the sign-in page, the token is read and sent to the` /api/auth/verify-token` endpoint for decoding and verification. This step works successfully, and the auth cookie is saved correctly. However, the `/api/auth/session` request still returns an empty session `{}` in production, even though everything works as expected in development.

### Expected behavior

In production, the /api/auth/session endpoint should return the proper session data as it does in development. The session callbacks should be triggered in production as they are in development.

**Comment by tayloraucoin** [comment]
For added context, here is the network request in question, where the sufficient request is made but only an empty response is returned. 

To emphasize, this request triggers the `jwt` and `session` callbacks in development, but does not in production. 

<img width="1005" alt="Screen Shot 2024-09-05 at 2 41 57 PM" src="https://github.com/user-attachments/assets/50da72e1-8b64-4fbc-a396-efd25f9bd07a">
<img width="1005" alt="Screen Shot 2024-09-05 at 2 42 08 PM" src="https://github.com/user-attachments/assets/524b84ef-5969-4917-8c27-ea4ab34a8f1a">
<img width="1006" alt="Screen Shot 2024-09-05 at 2 42 21 PM" src="https://github.com/user-attachments/assets/1b86047f-9f40-4a51-8321-ea30ec92713c">
<img width="1007" alt="Screen Shot 2024-09-05 at 2 42 31 PM" src="https://github.com/user-attachments/assets/b7af0adf-692b-423c-9354-b80bc443a166">

**Reply by zhuojg** [reply]
Did you set the `NEXTAUTH_URL` environment variable in your production deployment? I ran into the same issue, and adding this variable resolved it for me.

**Comment by tayloraucoin** [comment]
@balazsorban44 why was this converted into a discussion? This seems to be a legitimate issue.

**Comment by praveenkumar-TU** [comment]
any solutions so far?

**Reply by hitenbytes** [reply]
@tayloraucoin Did you resolve this issue? I have same issue
